cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/AbstractConnectOperator.java

```diff
@@ -198,96 +199,87 @@ public ConnectOperatorMetricsHolder metrics()   {
+     * @param onClose Callback called when the watch is closed.
      *
      * @return A future which completes when the watch has been set up.
      */
-    public static Future<Watch> createConnectorWatch(AbstractConnectOperator<KubernetesClient, KafkaConnect, KafkaConnectList, Resource<KafkaConnect>, KafkaConnectSpec, KafkaConnectStatus> connectOperator,
-                                                     String watchNamespaceOrWildcard, Labels selectorLabels) {
+    public Future<Watch> createConnectorWatch(String namespace, Consumer<WatcherException> onClose) {
+        Labels selectorLabels =  config.getCustomResourceSelector();
         Optional<LabelSelector> selector = (selectorLabels == null || selectorLabels.toMap().isEmpty()) ? Optional.empty() : Optional.of(new LabelSelector(null, selectorLabels.toMap()));
 
-        return VertxUtil.async(connectOperator.vertx, () -> {
-            Watch watch = connectOperator.connectorOperator.watch(watchNamespaceOrWildcard, new Watcher<>() {
-                @Override
-                public void eventReceived(Action action, KafkaConnector kafkaConnector) {
-                    String connectorName = kafkaConnector.getMetadata().getName();
-                    String connectorNamespace = kafkaConnector.getMetadata().getNamespace();
-                    String connectorKind = kafkaConnector.getKind();
-                    String connectName = kafkaConnector.getMetadata().getLabels() == null ? null : kafkaConnector.getMetadata().getLabels().get(Labels.STRIMZI_CLUSTER_LABEL);
-                    String connectNamespace = connectorNamespace;
-
-                    switch (action) {
-                        case ADDED:
-                        case DELETED:
-                        case MODIFIED:
-                            if (connectName != null) {
-                                // Check whether a KafkaConnect exists
-                                connectOperator.resourceOperator.getAsync(connectNamespace, connectName)
-                                        .compose(connect -> {
-                                            KafkaConnectApi apiClient = connectOperator.connectClientProvider.apply(connectOperator.vertx);
-                                            if (connect == null) {
-                                                Reconciliation r = new Reconciliation("connector-watch", connectOperator.kind(),
-                                                        kafkaConnector.getMetadata().getNamespace(), connectName);
-                                                updateStatus(r, noConnectCluster(connectNamespace, connectName), kafkaConnector, connectOperator.connectorOperator);
-                                                LOGGER.infoCr(r, "{} {} in namespace {} was {}, but Connect cluster {} does not exist", connectorKind, connectorName, connectorNamespace, action, connectName);
+        connectorWatcher =  new OperatorWatcher<KafkaConnector>(this, namespace, onClose) {
```

Comment:
Does it really make sense to use the OperatorWatcher? You basically throw away the whole logic it has. So IMHO this is a bad reuse. This deserves another logic or another level of abstraction there.

Reply:
I'm not really sure which part would you extract. Keep in mind that the logic is shared between Connect and MM2 and the connectors do not really have their own reconciliation. (Not saying that there is nothing to extract - just not sure it is obvious what to extract)

Reply:
> I think we could extract the KafkaConnector reconciliation logic and copy into its own KafkaConnectorAssemblyOperator

Was experimenting with this before, but reverted due to the things Jakub mentioned above. Mess but not impossible. Also had the idea of creating a `KafkaConnectorWatch` class like we have a `NodePoolWatch`.  I'll experiment with this a little more

Reply:
Yes, KafkaNodePoolWatcher seems a reasonable approach, because the relationship Kafka-NodePools and KafkaConnect-KafkaConnector are similar.

Reply:
Refactored the logic into a `KafkaConnectorWatcher` class, let me know what you think